src/mongo/db/exec/shard_name.cpp

```diff
@@ -0,0 +1,104 @@
+/**
```

Comment:
The design of the execution engine is to have as few PlanStages as possible for only the most fundamental query execution concepts. (On the other hand, the number of expressions and accumulators should be quite large and should grow to meet the needs of applications.) Therefore, the query team is reluctant to add a new PlanStage just for the purpose of attaching shard name metadata.

This brings up an interesting question of who should be responsible for attaching the shardName metadata. One idea is to attach this in the [ShardFilterStage](https://github.com/mongodb/mongo/blob/97ffc69bea04758e1c7f94e1075ef2bcce652999/src/mongo/db/exec/shard_filter.h#L72). We should only attach this metadata, however, if static analysis of the query shows that there is a subsequent $meta:"shardName" expression. There are some changes in the works as part of SERVER-43146 that should make it easier to determine whether the query has this "shardName" metadata dependency.